### PR TITLE
Create image gallery for reusable cover images

### PR DESCRIPTION
**Pull request by jonasdeluna**
# Description

Implement imagegallery for setting and using preset photos on events.

# Result
image

image

# Testing

- [x] I have thoroughly tested my changes.

relies on: https://github.com/webkom/lego/pull/3312
---

Resolves ABA-150

### CODE REVIEW

File: app/routes/events/components/ImageGallery.tsx [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  return (
+    <>
+      <Modal>
+        {imageGallery.map((e) => {
+          <Image></Image>;
+        })}
+      </Modal>
+    </>
+  );
+};
```
Comment (ivarnakken) [inline app/routes/events/components/ImageGallery.tsx:18]:
```suggestion
const ImageGalleryComponent = ({
  imageGallery,
}: {
  ImageGallery: imageGallery;
}) => (
    <Modal>
      {imageGallery.map((e) => {
        <Image></Image>;
      })}
    </Modal>
 );
```

File: app/actions/FileActions.ts [hunk 2]
```diff
@@ -78,3 +79,24 @@ export function uploadFile({
...
+  query?: Record<string, any>;
+  next?: boolean;
+} = {}): Thunk<any> {
+  return callAPI({
+    types: ImageGallery.FETCH_ALL,
+    endpoint: '/events/imagegallery/',
+    schema: [imageGallerySchema],
+    query,
+    method: 'GET',
+    json: true,
```
Comment (ivarnakken) [inline app/actions/FileActions.ts:95]: Is `json: true` not the default? Or why is this needed?

File: app/routes/events/components/ImageGallery.tsx [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import type { imageGallery } from 'app/models';
+
+const ImageGalleryComponent = ({
+  imageGallery,
+}: {
+  ImageGallery: imageGallery;
+}) => {
+  return (
+    <>
+      <Modal>
```
Comment (ivarnakken) [inline app/routes/events/components/ImageGallery.tsx:11]: In the image you showed, the modal has a title ("Bildegalleri:"). I don't see where it is set. I was going to say that it didn't need the semicolon and that it didn't need to be centered, so that it matches the other modals.

**Review by ivarnakken — COMMENTED** [review]
This was sort of a rushed review by me, sorry, but it looks overall very nice and clean! 💯 Very cool feature, so good job!

File: app/routes/events/components/ImageGallery.tsx (hunk 3 above)
Comment (jonasdeluna) [reply app/routes/events/components/ImageGallery.tsx:11]: Ok somehow this is just one of the WIP commits, not sure how but the rebase got messed up... I'll try to push the latest one tomorrow probably

**Comment by jonasdeluna** [comment]
I pushed the non-rebased thingy now, just need to rebase it and that's what I originally meant to push. I will look through the reviews that still apply though!

**Comment by jonasdeluna** [comment]
ABA-150

File: app/routes/events/components/EventEditor/EventEditor.css [hunk 4]
```diff
@@ -62,6 +62,42 @@
...
+  border-radius: var(--border-radius-md);
+}
+
+.imageGalleryEntry:hover {
+  opacity: 0.8;
+}
+
+.imageGalleryEntry:active {
+  opacity: 0.65;
+}
```
Comment (ivarnakken) [inline app/routes/events/components/EventEditor/EventEditor.css:83]:
```suggestion
.imageGalleryEntry {
  border-radius: var(--border-radius-md);
  
  &:hover {
    opacity: 0.8;
  }
  
  &:active {
    opacity: 0.65;
  }
}
```

File: app/components/Image/Image.tsx [hunk 5]
```diff
@@ -39,7 +47,11 @@ const ImageComponent = (props: Props) => {
     }
 
     const image = new Image();
-    image.src = src;
+    if (darkModeSource) {
+      image.src = darkModeSource;
+    } else {
+      image.src = src;
+    }
```
Comment (ivarnakken) [inline app/components/Image/Image.tsx:54]: Wait how is this `darkModeSource` relevant here in this PR? It's a nice feature, but I can't see the relevance. How does a user even upload a dark mode source?

File: app/routes/events/components/EventEditor/index.tsx [hunk 6]
```diff
@@ -127,10 +145,91 @@ function EventEditor({
...
+          <Tooltip content="Lagre bildet til bildegalleriet. Det kan da brukes i andre arrangementer.">
+            <Field
+              label="Lagre til bildegalleriet"
+              name="saveToImageGallery"
+              component={CheckBox.Field}
+              fieldClassName={styles.metaField}
+              className={styles.formField}
+              normalize={(v) => !!v}
+            />
+          </Tooltip>
```
Comment (ivarnakken) [inline app/routes/events/components/EventEditor/index.tsx:231]: Regarding ABA-494, it would be nice if the tooltip did not wrap the entire field.

File: app/routes/events/components/EventEditor/index.tsx [hunk 7]
```diff
@@ -127,10 +145,91 @@ function EventEditor({
...
+                          }}
+                          className={styles.imageGalleryEntry}
+                        />
+                      </div>
+                    );
+                  })}
+                {imageGallery.length === 0 && (
+                  <h2>
+                    Det finnes ingen bilder i bildegalleriet. Hvorfor ikke laste
+                    opp ett?
```
Comment (ivarnakken) [inline app/routes/events/components/EventEditor/index.tsx:194]:
```suggestion
                    opp et?
```

**Review by ivarnakken — CHANGES_REQUESTED** [review]
Would be awesome to get this in soon! 😄 You've done great work! 🥇 

I now realize I've made a lot of comments, but if it helps you may resolve some of them in a new PR - as long as you do hehe

File: app/actions/FileActions.ts [hunk 8]
```diff
@@ -1,6 +1,7 @@
 import slug from 'slugify';
+import { imageGallerySchema } from 'app/reducers';
 import type { Thunk } from 'app/types';
-import { File as FileType } from './ActionTypes';
+import { File, File as FileType, ImageGallery } from './ActionTypes';
```
Comment (eikhr) [inline app/actions/FileActions.ts:4]:
```suggestion
import { File as FileType, ImageGallery } from './ActionTypes';
```
I think the type `UploadArgs.file` which is set to `File` on line 42 is intended to be the built-in [DOM-File type](https://developer.mozilla.org/en-US/docs/Web/API/File). Importing `File` here overshadows this, which is probably why it was imported as `FileType` earlier. 

If you just remove `File` from the import and change the action-type: `types: File.PATCH` to `types: FileType.PATCH` on line 109, I believe all the types will be correct.

File: app/reducers/imageGallery.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
+import { produce } from 'immer';
+import { createSelector } from 'reselect';
+import createEntityReducer, { fetching } from 'app/utils/createEntityReducer';
+import { ImageGallery, File } from '../actions/ActionTypes';
+
+export default createEntityReducer({
+  key: 'imageGallery',
```
Comment (eikhr) [inline app/reducers/imageGallery.ts:7]:
```suggestion
  key: 'imageGalleryEntries',
```
The naming convention (which is aaalmost always followed) for entityReducer keys is the plural of the entity.

File: app/reducers/imageGallery.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+        }
+        break;
+    }
+    return state;
+  },
+  /*   schema: imageGallerySchema,
+  fetching: fetching(ImageGallery.FETCH_ALL),
+ */
+});
+export const selectImageGalleries = createSelector(
```
Comment (eikhr) [inline app/reducers/imageGallery.ts:27]:
```suggestion
export const selectImageGalleryEntries = createSelector(
```
Just to be consistent with our language here, I understand it as one image gallery with multiple images (or entries?) inside. 
It doesn't really matter what we call things, but it should be consistent.

File: app/routes/events/components/EventEditor/index.tsx [hunk 11]
```diff
@@ -127,10 +145,91 @@ function EventEditor({
...
-          edit={isEditPage && ((token) => setCoverPhoto(eventId, token))}
           aspectRatio={20 / 6}
-          img={event.cover}
+          img={useImageGallery ? imageGalleryUrl : event.cover}
         />
+        <Modal
+          show={showImageGallery}
+          onHide={() => setShowImageGallery(false)}
+          contentClassName={styles.imageGallery}
+        >
```
Comment (eikhr) [inline app/routes/events/components/EventEditor/index.tsx:155]: It might be an idea to put the modal (or just the content) into a separate component, if it's not too much work. This file is already huuge

File: app/routes/events/components/EventEditor/index.tsx [hunk 12]
```diff
@@ -562,41 +661,6 @@ function EventEditor({
...
-                </>
-              }
-              name="isClarified"
-              component={CheckBox.Field}
-              fieldClassName={styles.metaFieldInformation}
-              className={styles.formField}
-              normalize={(v) => !!v}
-            />
-          </Tooltip>
-        )}
```
Comment (eikhr) [inline app/routes/events/components/EventEditor/index.tsx:599]: Intentional?

**Review by eikhr — CHANGES_REQUESTED** [review]
I think you should take a quick look at the naming of stuff to make sure everything is consistent and makes sense. 
It looks like you've accidentally deleted the event-rules checkbox thing from the EventEditor.
Other than that, everything looks good!

File: app/routes/events/components/EventEditor/index.tsx [hunk 13]
```diff
@@ -83,8 +95,14 @@ function EventEditor({
...
   initialized,
   push,
+  imageGallery,
+  change,
+  setSaveForUse,
 }: Props) {
   const isEditPage = eventId !== undefined;
+  const [showImageGallery, setShowImageGallery] = useState(false);
+  const [useImageGallery, setUseImageGallery] = useState(false);
+  const [imageGalleryUrl, setImageGalleryUrl] = useState('');
```
[the remark this answers is not in the document]
Comment (jonasdeluna) [reply app/routes/events/components/EventEditor/index.tsx:107]: Not sure if such an image exists and is guaranteed to exist in the db?

File: app/components/Image/Image.tsx (hunk 5 above)
Comment (jonasdeluna) [reply app/components/Image/Image.tsx:59]: This is for static images from /public I believe and just snuck in. There's another issue on dynamic uploadable dark mode images.

File: app/routes/events/components/EventEditor/index.tsx [hunk 14]
```diff
@@ -675,10 +739,6 @@ const validate = (data) => {
     errors.feedbackDescription = 'Kan ikke være tomt';
   }
 
-  if (!data.isClarified) {
-    errors.isClarified = 'Arrangementet må være avklart';
-  }
```
[the remark this answers is not in the document]
Comment (jonasdeluna) [reply app/routes/events/components/EventEditor/index.tsx:672]: Oops fixed now

File: app/models.ts [hunk 15]
```diff
@@ -146,10 +146,16 @@ export type Group = {
...
+type Cover = {
+  url: string;
+  fileKey: string;
+};
+
 type EventBase = {
   id: ID;
   title: string;
-  cover: string;
+  cover: Cover;
```
Comment (ivarnakken) [inline app/models.ts:158]: Could you update the model in `app/store/models/Event.d.ts` as well? 😄

File: app/reducers/imageGallery.ts [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+            draft.items = draft.items.filter((id) => id !== action.meta.id);
+          });
+        }
+        break;
+    }
+    return state;
+  },
+  /*   schema: imageGallerySchema,
+  fetching: fetching(ImageGallery.FETCH_ALL),
+ */
```
[the remark this answers is not in the document]
Comment (ivarnakken) [reply app/reducers/imageGallery.ts:25]: ^

File: app/routes/events/components/EventEditor/index.tsx [hunk 17]
```diff
@@ -127,10 +145,91 @@ function EventEditor({
...
+            <Flex column>
+              <h1>Bildegalleri</h1>
+              <Flex wrap alignItems="center" justifyContent="space-around">
+                {imageGallery &&
+                  imageGallery.map((e) => {
+                    return (
+                      <div key={e.key} className={styles.imageGalleryContainer}>
+                        <ConfirmModalWithParent
+                          title="Fjern fra galleri"
+                          message="Er du sikker på at du vil fjerne bildet fra bildegalleriet? (Bildet blir ikke slettet!)"
```
[the remark this answers is not in the document]
Comment (ivarnakken) [reply app/routes/events/components/EventEditor/index.tsx:168]: ^

File: app/routes/events/components/EventEditor/index.tsx [hunk 18]
```diff
@@ -127,10 +145,91 @@ function EventEditor({
...
+                          onConfirm={() =>
+                            setSaveForUse(`${e.key}:${e.token}`, false)
```
[the remark this answers is not in the document]
Comment (ivarnakken) [reply app/routes/events/components/EventEditor/index.tsx:170]: ^

File: app/routes/events/components/EventEditor/index.tsx [hunk 19]
```diff
@@ -125,10 +147,91 @@ function EventEditor({
...
+        >
+          <>
+            <Flex column>
+              <h1>Bildegalleri</h1>
+              <Flex wrap alignItems="center" justifyContent="space-around">
+                {imageGallery &&
+                  imageGallery.map((e) => {
+                    return (
+                      <div key={e.key} className={styles.imageGalleryContainer}>
+                        <ConfirmModalWithParent
```
Comment (ivarnakken) [inline app/routes/events/components/EventEditor/index.tsx:166]: The confirm modal is currently implemented differently. Take a look at use cases of `ConfirmModal`

File: app/routes/events/components/EventEditor/index.tsx [hunk 20]
```diff
@@ -125,10 +147,91 @@ function EventEditor({
...
+                          title="Fjern fra galleri"
+                          message="Er du sikker på at du vil fjerne bildet fra bildegalleriet? (Bildet blir ikke slettet!)"
+                          onConfirm={() =>
+                            setSaveForUse(`${e.key}:${e.token}`, false)
+                          }
+                          closeOnConfirm
+                        >
+                          <button className={styles.closeButton}>
+                            <Icon name="close-circle" size={24} />
+                          </button>
```
Comment (ivarnakken) [inline app/routes/events/components/EventEditor/index.tsx:176]: You can move the onclick function from ConfirmModal into the Icon component and it will resemble all other icons instead of a custom closeButton style

File: app/routes/events/components/EventEditor/index.tsx (hunk 7 above)
Comment (ivarnakken) [reply app/routes/events/components/EventEditor/index.tsx:196]: ^

File: app/routes/events/components/EventEditor/index.tsx [hunk 21]
```diff
@@ -127,10 +145,91 @@ function EventEditor({
...
+              onClick={() => setShowImageGallery(true)}
+            >
+              Velg bilde fra Bildegalleriet
+            </Button>
+            <Tooltip content="Velg et bilde som er lastet opp og markert for bruk fra tidligere arrangementer.">
+              <Icon
+                name="information-circle-outline"
+                size={20}
+                style={{
+                  cursor: 'pointer',
```
[the remark this answers is not in the document]
Comment (ivarnakken) [reply app/routes/events/components/EventEditor/index.tsx:217]: ^ please

File: app/routes/events/components/EventEditor/index.tsx [hunk 22]
```diff
@@ -127,10 +145,91 @@ function EventEditor({
...
+          </>
+        </Modal>
+
+        <Flex wrap alignItems="center" justifyContent="space-between">
+          <Flex alignItems="center" justifyContent="space-between">
+            <Button
+              className={styles.scannerButton}
+              onClick={() => setShowImageGallery(true)}
+            >
+              Velg bilde fra Bildegalleriet
```
[the remark this answers is not in the document]
Comment (ivarnakken) [reply app/routes/events/components/EventEditor/index.tsx:210]: ^

File: app/routes/events/components/EventEditor/index.tsx [hunk 23]
```diff
@@ -127,10 +145,91 @@ function EventEditor({
...
+                )}
+              </Flex>
+            </Flex>
+          </>
+        </Modal>
+
+        <Flex wrap alignItems="center" justifyContent="space-between">
+          <Flex alignItems="center" justifyContent="space-between">
+            <Button
+              className={styles.scannerButton}
```
[the remark this answers is not in the document]
Comment (ivarnakken) [reply app/routes/events/components/EventEditor/index.tsx:207]: ^

File: app/routes/events/components/EventEditor/EventEditor.css [hunk 24]
```diff
@@ -61,6 +61,42 @@
 
 .subSection {
   margin-left: 10px;
-  border-left: 4px solid var(--lego-red-color);
+  border-left: 4px solid var(--danger-color);
```
Comment (ivarnakken) [inline app/routes/events/components/EventEditor/EventEditor.css:64]: Why is this a danger color?

**Comment by jonasdeluna** [comment]
Okay I believe I fixed everything except the string splitting and the hover wrap around.

File: app/models.ts [hunk 25]
```diff
@@ -233,6 +239,15 @@ export type EventPool = EventPoolBase & {
...
 };
 
+type ImageGalleryEntry = {
+  key: string;
+  cover: string;
+  token: string;
+  coverPlaceholder: string;
+};
+
+export type imageGallery = ImageGalleryEntry[];
```
Comment (ivarnakken) [inline app/models.ts:249]: Capital letters on types

File: app/routes/events/components/EventEditor/index.tsx [hunk 26]
```diff
@@ -125,10 +144,90 @@ function EventEditor({
...
+                      }
+                    >
+                      {({ openConfirmModal }) => (
+                        <button
+                          className={styles.closeButton}
+                          onClick={openConfirmModal}
+                        >
+                          <Icon name="close-circle" />
+                        </button>
+                      )}
```
Comment (ivarnakken) [inline app/routes/events/components/EventEditor/index.tsx:176]:
```suggestion
                      {({ openConfirmModal }) => (
                          <Icon onClick={openConfirmModal} name="close-circle" />
                      )}
```

File: app/routes/events/components/EventEditor/index.tsx [hunk 27]
```diff
@@ -125,10 +144,90 @@ function EventEditor({
...
+            <Button onClick={() => setShowImageGallery(true)}>
+              Velg bilde fra bildegalleriet
+            </Button>
+            <Tooltip content="Velg et bilde som er lastet opp og markert for bruk fra tidligere arrangementer.">
+              <Icon
+                name="information-circle-outline"
+                size={20}
+                style={{
+                  cursor: 'help',
+                  marginLeft: '0.5em',
```
Comment (ivarnakken) [inline app/routes/events/components/EventEditor/index.tsx:214]: Is the left margin still needed? With the recent changes to the tooltip it should not

File: app/routes/events/components/EventEditor/index.tsx [hunk 28]
```diff
@@ -620,15 +719,15 @@ const validate = (data) => {
   }
 
   if (!data.title) {
-    errors.title = 'Tittel er påkrevd';
+    errors.title = 'Tittel er påkrevet';
```
Comment (ivarnakken) [inline app/routes/events/components/EventEditor/index.tsx:722]: https://naob.no/ordbok/p%C3%A5krevd

Why are these changed

File: app/routes/events/components/EventEditor/EventEditor.css [hunk 29]
```diff
@@ -64,3 +64,41 @@
...
+
+.imageGallery {
+  width: 60%;
+}
+
+.imageGalleryEntry {
+  border-radius: var(--border-radius-md);
+  &:hover{
+    opacity: 0.8;
+
```
Comment (ivarnakken) [inline app/routes/events/components/EventEditor/EventEditor.css:76]: Whitespace

File: app/routes/events/components/EventEditor/EventEditor.css [hunk 30]
```diff
@@ -64,3 +64,41 @@
...
+  }
+
+  &:active{
+    opacity: 0.65;
+
```
Comment (ivarnakken) [inline app/routes/events/components/EventEditor/EventEditor.css:81]: More whitespace

File: app/routes/events/components/EventEditor/index.tsx (hunk 27 above)
Comment (jonasdeluna) [reply app/routes/events/components/EventEditor/index.tsx:214]: It looks a little wack without actually, I'll leave it for now (?)

**Comment by jonasdeluna** [comment]
I'm not sure what you meant by the tooltip thingy or how to make it go to the right.
As for the string value it's just the value of the cover form value. I moved the string split stuff to only occur in one place and it's a little cleaner. Aside from that and eventually moving it into a it's own component it would be nice to merge this soon...
